File: backend/engine/dice.py
"""
engine/dice.py
--------------
Dice rolling utilities. Pure Python, no external dependencies.
Supports expressions: "4", "D6", "D6+2", "2D3", "2D6+1"
"""

import random
import re


def roll(expr: str | int) -> int:
    """Roll a dice expression, return integer result (minimum 1)."""
    if isinstance(expr, (int, float)):
        return max(1, int(expr))

    s = str(expr).upper().strip()

    if s.isdigit():
        return max(1, int(s))

    # D6
    m = re.fullmatch(r"D(\d+)", s)
    if m:
        return random.randint(1, int(m.group(1)))

    # D6+2
    m = re.fullmatch(r"D(\d+)\+(\d+)", s)
    if m:
        return random.randint(1, int(m.group(1))) + int(m.group(2))

    # 2D6
    m = re.fullmatch(r"(\d+)D(\d+)", s)
    if m:
        n, f = int(m.group(1)), int(m.group(2))
        return sum(random.randint(1, f) for _ in range(n))

    # 2D6+1
    m = re.fullmatch(r"(\d+)D(\d+)\+(\d+)", s)
    if m:
        n, f, b = int(m.group(1)), int(m.group(2)), int(m.group(3))
        return sum(random.randint(1, f) for _ in range(n)) + b

    return 1  # fallback


def clamp(v: int, lo: int, hi: int) -> int:
    return max(lo, min(hi, v))


def wound_threshold(strength: int, toughness: int) -> int:
    """Return the minimum die result needed to wound (WH40K 10e rules)."""
    if strength >= 2 * toughness:
        return 2
    elif strength > toughness:
        return 3
    elif strength == toughness:
        return 4
    elif toughness >= 2 * strength:
        return 6
    else:
        return 5

File: frontend/src/engine/dice.js
/**
 * engine/dice.js
 * Dice rolling utilities — port of backend/engine/dice.py
 * Supports: "4", "D6", "D6+2", "2D3", "2D6+1"
 */

export function roll(expr) {
  if (typeof expr === 'number') return Math.max(1, Math.floor(expr))

  const s = String(expr).toUpperCase().trim()

  if (/^\d+$/.test(s)) return Math.max(1, parseInt(s, 10))

  let m

  // D6
  m = s.match(/^D(\d+)$/)
  if (m) return randInt(1, parseInt(m[1]))

  // D6+2
  m = s.match(/^D(\d+)\+(\d+)$/)
  if (m) return randInt(1, parseInt(m[1])) + parseInt(m[2])

  // 2D6
  m = s.match(/^(\d+)D(\d+)$/)
  if (m) {
    const [, n, f] = m
    return sumDice(parseInt(n), parseInt(f))
  }

  // 2D6+1
  m = s.match(/^(\d+)D(\d+)\+(\d+)$/)
  if (m) {
    const [, n, f, b] = m
    return sumDice(parseInt(n), parseInt(f)) + parseInt(b)
  }

  return 1 // fallback
}

function randInt(min, max) {
  return Math.floor(Math.random() * (max - min + 1)) + min
}

function sumDice(n, faces) {
  let total = 0
  for (let i = 0; i < n; i++) total += randInt(1, faces)
  return total
}

export function clamp(v, lo, hi) {
  return Math.max(lo, Math.min(hi, v))
}

export function woundThreshold(strength, toughness) {
  if (strength >= 2 * toughness) return 2
  if (strength > toughness)      return 3
  if (strength === toughness)    return 4
  if (toughness >= 2 * strength) return 6
  return 5
}

export function d6() {
  return randInt(1, 6)
}
